test/app-luatest/socket_test.lua

```diff
@@ -35,3 +35,16 @@ g.test_from_fd = function()
     t.assert(s.itype, 0)
     t.assert_not(s:close())
 end
+
+g.test_detach = function()
+    local s1 = socket('AF_INET', 'SOCK_STREAM', 'tcp')
+    t.assert(s1)
+    local s2 = socket.from_fd(s1:fd())
+    t.assert(s2)
+    t.assert_error_msg_content_equals('Usage: socket:method()', s1.detach)
+    t.assert_is(s1:detach(), nil)
```

Comment:
Shouldn't we loss the `s1` object, call `collectgarbage()` and ensure that `s2` remains valid?

Reply:
Makes sense. Added a test case:

```diff
diff --git a/test/app-luatest/socket_test.lua b/test/app-luatest/socket_test.lua
index 701b5f6bc389..08dbc40ecbcc 100644
--- a/test/app-luatest/socket_test.lua
+++ b/test/app-luatest/socket_test.lua
@@ -48,6 +48,16 @@ g.test_detach = function()
     t.assert_error_msg_content_equals(errmsg, s1.detach, s1)
     t.assert_error_msg_content_equals(errmsg, s1.close, s1)
     t.assert(s2:close())
+
+    s1 = socket('AF_UNIX', 'SOCK_STREAM', 0)
+    t.assert(s1)
+    s2 = socket.from_fd(s1:fd())
+    t.assert(s2)
+    t.assert_is(s1:detach())
+    s1 = nil
+    collectgarbage('collect')
+    t.assert(s2:readable())
+    t.assert(s2:close())
 end
 
 g.test_socketpair = function()
```